blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+2. `minor` version if backward **source compatibility** is broken, and
+3. `patch` version to signal neither binary nor source incompatibility.
+
+- When the `major` version is `0`, a minor version increment MAY contain **both source and binary breakages**, but a patch version increment MUST remain **binary compatible**.
+
+We call this Early SemVer, because according to the [Semantic Versioning Spec][2] there are no guarantees between any versions when the major version is `0`. In the Scala library ecosystem, though, we often start guaranteeing binary compatibility for `0.y.z` like sbt 0.13 and Scala.JS 0.6.
+
+Unfortunately, it is not an easy task to know whether a change on public API broke source or binary compatibility. Scala Center contracted Alexandre Archambault to create [sbt-version-policy][5]. This plugin helps library authors to self-check a version scheme. To use this, add the following to your `project/plugins.sbt`:
+
+~~~ scala
+addSbtPlugin("ch.epfl.scala" % "sbt-version-policy" % "1.0.0-RC5")
+~~~
+
+And declare your compatibility intention regarding the next release, in your `build.sbt`:
+
+~~~
+// The next release must be binary and source compatible with the previous release (it will be a patch release)
+ThisBuild / versionPolicyIntention := Compatibility.BinaryAndSourceCompatible
+~~~
+
+This plugin provides a task called `versionPolicyCheck`, which you can call in the CI (continuous integration) server. The task performs an automatic binary compatibility check.
+
+~~~ text
+> versionPolicyCheck
+~~~
+
+I think Early Semver gives better flexibility to both library authors and library users since it gives more information about what would be in minor upgrades (I often call this feature release) vs patch. For example, sbt 1 ships bug fixes as 1.3.x patch releases without going through RC cycle, so bug fixes are released quickly.
+
+During the minor upgrade (feature release), we aggressively add new features, while maintaining binary compatibility for the plugin ecosystem. These only come out once a year, and it goes through the RC cycle.
+
+## versionScheme, libraryDependencySchemes, and sbt 1.5.0
+
+As mentioned earlier, sbt contains a built-in eviction warning feature, but there are too many false positives right now because it needs to guess.
+
+Since sbt 1.4.0, there is a new setting `versionScheme`, which can be used by library authors as follows in `build.sbt`:
+
+~~~ scala
+ThisBuild / versionScheme := Some("early-semver")
+~~~
+
+sbt 1.4.0 includes this information into `pom.xml` and `ivy.xml` as a property. In addition, sbt uses the information to take the guessing out of eviction warning when this information is available.
+
+In sbt 1.5.0, eviction warnings will be replaced with [eviction errors][8]. Since we can now reliably detect whether two dependencies with different versions are compatible, or if they conflict, the build will fail if an incompatibility is detected in your dependencies.
```

Comment:
What happens for libraries that did not configure their `versionScheme`. Am I going to suddenly get errors when upgrading to sbt 1.5.0? What can I do as a user to prevent that? How do I migrate, concretely?

Reply:
> What happens for libraries that did not configure their `versionScheme`. Am I going to suddenly get errors when upgrading to sbt 1.5.0?

No, you won’t get any errors. In sbt 1.5.0, libraries that don’t configure their `versionScheme` will always be silently evicted in case of conflict. And, by the way, I think this is not a good thing. There is a discussion about that in https://github.com/sbt/sbt/pull/6221#issuecomment-748821082.

Reply:
Eugene is working on restoring the old behavior for libraries that don’t provide their `versionScheme`, so I’m marking this discussion as “resolved”.